src/os/bluestore/BlueFS.cc

```diff
@@ -2546,91 +2633,101 @@ void BlueFS::_compact_log_async(std::unique_lock<ceph::mutex>& l)
-      discarded += drop;
-      dout(10) << __func__ << "   kept " << e << " removed " << temp << dendl;
-    }
-    old_extents.push_back(temp);
-  }
-  auto from = log_file->fnode.extents.begin();
-  auto to = log_file->fnode.extents.end();
-  while (from != to) {
-    new_log->fnode.append_extent(*from);
-    ++from;
+      // keep, but how much?
+      if (processed < old_log_jump_to) {
+	ceph_assert(processed + e.length > old_log_jump_to);
+	ceph_assert(old_log_jump_to - processed <= std::numeric_limits<uint32_t>::max());
+	uint32_t cut_at = uint32_t(old_log_jump_to - processed);
+	// need to cut, first half gets dropped
+	bluefs_extent_t retire(e.bdev, e.offset, cut_at);
+	old_extents.push_back(retire);
+	// second half goes to new log
+	bluefs_extent_t keep(e.bdev, e.offset + cut_at, e.length - cut_at);
+	new_log->fnode.append_extent(keep);
+	dout(10) << __func__ << " kept " << keep << " removed " << retire << dendl;
+      } else {
+	// take entire extent
+	ceph_assert(processed >= old_log_jump_to);
+	new_log->fnode.append_extent(e);
+	dout(10) << __func__ << " kept " << e << dendl;
+      }
+    }
+    processed += e.length;
   }
   // we will write it to super
   new_log->fnode.reset_delta();
 
+  // 6. write the super block to reflect the changes
+  dout(10) << __func__ << " writing super" << dendl;
+  new_log->fnode.ino = log_file->fnode.ino;
+  new_log->fnode.size = 0;
+  new_log->fnode.mtime = ceph_clock_now();
+  super.log_fnode = new_log->fnode;
+  ++super.version;
+  _write_super(BDEV_DB);
+  _flush_bdev();
+
+  log.lock.lock();
+  // swapping log_file and new_log
   vselector->sub_usage(log_file->vselector_hint, log_file->fnode);
 
   // clear the extents from old log file, they are added to new log
   log_file->fnode.clear_extents();
   // swap the log files. New log file is the log file now.
   new_log->fnode.swap_extents(log_file->fnode);
 
-  log_writer->pos = log_writer->file->fnode.size =
-    log_writer->pos - old_log_jump_to + new_log_jump_to;
+  log.writer->pos = log.writer->file->fnode.size =
+    log.writer->pos - old_log_jump_to + new_log_jump_to;
 
   vselector->add_usage(log_file->vselector_hint, log_file->fnode);
 
-  // 6. write the super block to reflect the changes
-  dout(10) << __func__ << " writing super" << dendl;
-  super.log_fnode = log_file->fnode;
-  ++super.version;
-  _write_super(BDEV_DB);
+  log.lock.unlock();
 
-  lock.unlock();
-  flush_bdev();
-  lock.lock();
+  old_forbidden = atomic_exchange(&log_forbidden_to_expand, false);
+  ceph_assert(old_forbidden == true);
+  //to wake up if someone was in need of expanding log
+  log_cond.notify_all();
 
   // 7. release old space
   dout(10) << __func__ << " release old log extents " << old_extents << dendl;
-  for (auto& r : old_extents) {
-    pending_release[r.bdev].insert(r.offset, r.length);
+  {
+    std::lock_guard dl(dirty.lock);
+    for (auto& r : old_extents) {
+      pending_release[r.bdev].insert(r.offset, r.length);
```

Comment:
shouldn't pending_release go to dirty struct as it's protected with dirty.lock?

Reply:
This is a very good idea.
It allows to detect, that pending_release in `BlueFS::open_for_write()` is not protected.
+1